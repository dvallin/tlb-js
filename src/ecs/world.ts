import { Storage } from './storage'
import { System } from './system'
import { Resource } from './resource'
import { Entity, EntityModifier } from './entity'

export class World<C, S, R> {
  public readonly components: Map<C, Storage<object>> = new Map()
  public readonly resources: Map<R, Resource<C, S, R>> = new Map()
  public readonly systems: Map<S, System<C, S, R>> = new Map()
  public readonly emptySystems: Set<S> = new Set()
  public readonly activeSystems: Set<S> = new Set()

  private openEntities: Set<Entity> = new Set()
  private lastEntity: Entity = -1

  public get entityCount(): number {
    return this.lastEntity + 1 - this.openEntities.size
  }

  public registerComponentStorage<T extends object>(component: C, storage: Storage<T>): void {
    this.components.set(component, storage)
  }

  public getStorage<T extends object>(component: C): Storage<T> {
    return this.components.get(component)! as Storage<T>
  }

  public getResource<T extends Resource<C, S, R>>(resource: R): T {
    return this.resources.get(resource)! as T
  }

  public getComponent<T extends object>(entity: Entity, component: C): T | undefined {
    const storage = this.getStorage<T>(component)
    if (storage !== undefined) {
      return storage.get(entity)
    }
    return undefined
  }

  public hasEntity(entity: Entity): boolean {
    return entity <= this.lastEntity && !this.openEntities.has(entity)
  }

  public hasComponent(entity: Entity, component: C): boolean {
    const storage = this.getStorage(component)
    if (storage !== undefined) {
      return storage.has(entity)
    }
    return false
  }

  public createEntity(): EntityModifier<C, S, R> {
    let entity
    if (this.openEntities.size > 0) {
      entity = this.openEntities.values().next().value
      this.openEntities.delete(entity)
    } else {
      entity = ++this.lastEntity
    }
    return new EntityModifier(this, entity)
  }

  public editEntity(entity: Entity): EntityModifier<C, S, R> {
    return new EntityModifier(this, entity)
  }

  public deleteEntity(entity: Entity): void {
    this.editEntity(entity).delete()
    this.openEntities.add(entity)
  }

  public registerSystem(name: S, system: System<C, S, R>): void {
    this.systems.set(name, system)
  }

  public enableSystem(name: S): void {
    this.activeSystems.add(name)
  }

  public disableSystem(name: S): void {
    this.activeSystems.delete(name)
  }

  public activeSystemsList(): S[] {
    const l: S[] = []
    this.activeSystems.forEach(s => l.push(s))
    return l
  }

  public registerResource(resource: Resource<C, S, R>): void {
    this.resources.set(resource.kind, resource)
  }

  public updateResources(): void {
    for (const resource of this.resources) {
      resource[1].update(this)
    }
  }

  public updateSystems(): void {
    this.emptySystems.clear()
    for (const name of this.activeSystems) {
      const system = this.systems.get(name)!
      const firstStorage = this.getStorage(system.components[0])
      let calls = 0
      firstStorage.foreach((entity: Entity) => {
        let callSystem = true
        for (let i = 1; i < system.components.length; i++) {
          if (!this.hasComponent(entity, system.components[i])) {
            callSystem = false
            break
          }
        }
        if (callSystem) {
          system.update(this, entity)
          calls++
        }
      })
      if (calls === 0) {
        this.emptySystems.add(name)
      }
    }
  }
}
