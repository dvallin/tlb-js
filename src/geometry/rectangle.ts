import { AbstractShape } from "./shape"
import { Vector } from "../spatial"

export class Rectangle extends AbstractShape {

    public static fromBounds(left: number, right: number, top: number, bottom: number): Rectangle {
        return new Rectangle(left, top, right - left + 1, bottom - top + 1)
    }

    public constructor(
        public readonly x: number,
        public readonly y: number,
        public readonly width: number,
        public readonly height: number,
    ) {
        super()
    }

    public get left(): number {
        return this.x
    }

    public get right(): number {
        return this.x + this.width - 1
    }

    public get top(): number {
        return this.y
    }

    public get bottom(): number {
        return this.y + this.height - 1
    }

    public get topLeft(): Vector {
        return new Vector(this.left, this.top)
    }

    public get topRight(): Vector {
        return new Vector(this.right, this.top)
    }

    public get bottomRight(): Vector {
        return new Vector(this.right, this.bottom)
    }

    public get bottomLeft(): Vector {
        return new Vector(this.left, this.bottom)
    }

    public get center(): Vector {
        const cx = (this.right + this.left) / 2
        const cy = (this.bottom + this.top) / 2
        return new Vector(Math.floor(cx), Math.floor(cy))
    }

    public plus(other: Rectangle): Rectangle {
        return Rectangle.fromBounds(
            Math.min(this.left, other.left), Math.max(this.right, other.right),
            Math.min(this.top, other.top), Math.max(this.bottom, other.bottom)
        )
    }

    public bounds(): Rectangle {
        return this
    }

    public containsVector(p: Vector): boolean {
        return p.x >= this.x && p.x < this.x + this.width
            && p.y >= this.y && p.y < this.y + this.height
    }

    public translate(t: Vector): Rectangle {
        return new Rectangle(this.x + t.x, this.y + t.y, this.width, this.height)
    }

    public grow(): Rectangle {
        return new Rectangle(this.x - 1, this.y - 1, this.width + 2, this.height + 2)
    }

    public all(f: (position: Vector) => boolean): boolean {
        for (let i = 0; i < this.height; i++) {
            for (let j = 0; j < this.width; j++) {
                const position = new Vector(j + this.x, i + this.y)
                if (!f(position)) {
                    return false
                }
            }
        }
        return true
    }
}
