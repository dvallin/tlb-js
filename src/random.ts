import * as seedrandom from "seedrandom"

let rng: seedrandom.prng

export function init(seed?: string): void {
    rng = seedrandom(seed)
}

export function uniformInteger(min: number = 0, max: number = 1): number {
    return (Math.abs(rng.int32()) % (max - min)) + min
}

export function bernoulli(p: number = 0.5): boolean {
    return rng.double() <= p
}

export function binomialNormed(p: number[]): number {
    let test = rng.double()
    return p.findIndex((t) => {
        if (test <= t) {
            return true
        }
        test -= t
        return false
    })
}

export function dualDecision(
    p00: number, [p01, p10]: number[], p11: number
): [boolean, boolean] {
    const decision = binomialNormed([p00, p01, p10, p11])
    return [
        Math.floor(decision / 2) % 2 === 1,
        decision % 2 === 1
    ]
}

export function ternaryDecision(
    p000: number, [p001, p010, p100]: number[], [p011, p101, p110]: number[], p111: number
): [boolean, boolean, boolean] {
    const decision = binomialNormed([p000, p001, p010, p011, p100, p101, p110, p111])
    return [
        Math.floor(decision / 4) % 2 === 1,
        Math.floor(decision / 2) % 2 === 1,
        decision % 2 === 1
    ]
}
