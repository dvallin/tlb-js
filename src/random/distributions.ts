import { create, RandomSeed } from "random-seed"

export interface Distribution {
    sample(): number
}

export class Uniform implements Distribution {
    public readonly rng: RandomSeed

    public constructor(seed: string) {
        this.rng = create(seed)
    }

    public sample(): number {
        return this.rng.random()
    }
}

export class Exponential implements Distribution {

    public static fromSeed(seed: string, lambda: number = 1) {
        return new Exponential(new Uniform(seed), lambda)
    }

    private scale: number

    public constructor(
        public readonly distribution: Distribution,
        public readonly lambda: number = 1
    ) {
        this.scale = 1 - Math.exp(-this.lambda)
    }

    public sample(): number {
        const u = this.distribution.sample()
        return -Math.log(1 - this.scale * u) / this.lambda
    }
}
