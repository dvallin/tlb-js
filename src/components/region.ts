import { Shape } from '../geometry/shape'
import { Vector } from '../spatial'
import { Entity } from '../ecs/entity'

export type RegionsType = 'elevator' | 'red'

export interface RegionComponent {
  type: RegionsType
  level: number
  shape: Shape
  exits: Vector[]
  authorized: Set<Entity>
}

export type StructureType = 'room' | 'hub' | 'corridor'

export interface StructureComponent {
  region: Entity
  kind: StructureType
  shape: Shape
  connections: Entity[]
}
