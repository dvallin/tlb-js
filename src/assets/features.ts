import { Feature } from '../components/feature'
import { gray, primary } from '../renderer/palettes'
import { strangeSymbols, gridSymbols, arrows } from '../symbols'
import { Color } from '../renderer/color'

function character(character: string, name: string, diffuse: Color): Feature {
  return {
    cover: 'none',
    blocking: true,
    lightBlocking: true,
    ground: false,

    character,
    diffuse,
    name,
  }
}

function ground(name: string, character: string, diffuse: Color): Feature {
  return {
    cover: 'none',
    blocking: false,
    lightBlocking: false,
    ground: true,

    character,
    diffuse,
    name,
  }
}

function highObstacle(name: string, character: string, diffuse: Color): Feature {
  return {
    cover: 'full',
    blocking: true,
    lightBlocking: true,
    ground: false,

    character,
    diffuse,
    name,
  }
}

function lowObstacle(name: string, character: string, diffuse: Color): Feature {
  return {
    cover: 'partial',
    blocking: true,
    lightBlocking: false,
    ground: false,

    character,
    diffuse,
    name,
  }
}

function decoration(name: string, character: string, diffuse: Color): Feature {
  return {
    cover: 'none',
    blocking: false,
    lightBlocking: false,
    ground: false,

    character,
    diffuse,
    name,
  }
}

const featuresDefinition = {
  wall: highObstacle('wall', '#', gray[3]),
  corridor: ground('corridor', '.', gray[0]),
  room: ground('floor', '.', primary[1]),
  hub: ground('floor', '.', primary[0]),
  locker: highObstacle('locker', strangeSymbols[16], primary[1]),
  trash: decoration('trash', strangeSymbols[21], gray[2]),
  door: highObstacle('door', strangeSymbols[27], primary[1]),
  player: character('@', 'you', primary[0]),
  loot: decoration('some loot', 'l', primary[1]),
  table: lowObstacle('a table', strangeSymbols[3], primary[1]),
  civilian: character('c', 'civilian', primary[1]),
  guard: character('g', 'guard', primary[1]),
  eliteGuard: character('g', 'elite guard', primary[3]),
  terminal: lowObstacle('a terminal', strangeSymbols[28], primary[1]),
}
export type FeatureType = keyof typeof featuresDefinition
export const features: { [key in FeatureType]: Feature } = featuresDefinition

const generatorsDefinition: { [key: string]: (index: number) => Feature } = {
  block: (index: number) => highObstacle('a block of concrete', gridSymbols[[7, 5, 15, 17][index]], primary[1]),
  elevator: (_index: number) => lowObstacle('an elevator', arrows[7], primary[1]),
}
export type FeatureGeneratorsType = keyof typeof generatorsDefinition
export const generators: { [key in FeatureGeneratorsType]: (index: number) => Feature } = generatorsDefinition
