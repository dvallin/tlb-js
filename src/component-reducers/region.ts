import { TlbWorld } from '../tlb'
import { StructureComponent, RegionComponent } from '../components/region'
import { Entity } from '../ecs/entity'
import { PositionComponent } from '../components/position'
import { WorldMap, WorldMapResource } from '../resources/world-map'

export function isAuthorized(world: TlbWorld, position: PositionComponent, entity: Entity): boolean {
  const map: WorldMap = world.getResource<WorldMapResource>('map')
  const structure = map.levels[position.level].getStructure(position.position)!
  const room: StructureComponent = world.getComponent<StructureComponent>(structure, 'structure')!
  const region: RegionComponent = world.getComponent<RegionComponent>(room.region, 'region')!
  return region.authorized.has(entity)
}

export function authorize(world: TlbWorld, position: PositionComponent, entity: Entity): void {
  const map: WorldMap = world.getResource<WorldMapResource>('map')
  const structure = map.levels[position.level].getStructure(position.position)!
  const room: StructureComponent = world.getComponent<StructureComponent>(structure, 'structure')!
  const region: RegionComponent = world.getComponent<RegionComponent>(room.region, 'region')!
  region.authorized.add(entity)
}
