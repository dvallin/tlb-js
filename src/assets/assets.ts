import { Asset, Cover } from '../components/asset'
import { Vector } from '../spatial'
import { features, generators } from './features'
import { Feature } from '../components/feature'
import { DialogType } from './dialogs'

const s1x1 = new Vector([1, 1])
const s1x2 = new Vector([1, 2])
const s3x1 = new Vector([3, 1])
const s2x2 = new Vector([2, 2])
function asset(name: string, size: Vector, cover: Cover, dialog: DialogType | undefined, feature: (index: number) => Feature): Asset {
  return { name, size, cover, hasInventory: true, dialog, feature }
}

const assetsDefinition = {
  door: asset('metal door', s3x1, 'full', undefined, () => features['door']),
  locker: asset('a locker', s1x1, 'full', undefined, () => features['locker']),
  trash: asset('some trash', s1x1, 'none', undefined, () => features['trash']),
  loot: asset('a dead body', s1x1, 'none', undefined, () => features['loot']),
  table: asset('a table', s1x2, 'partial', undefined, () => features['table']),
  generator: asset('a generator', s2x2, 'full', undefined, generators['block']),
  elevator: asset('an elevator', s1x1, 'none', 'elevator', generators['elevator']),
}

export type AssetType = keyof typeof assetsDefinition
export const assets: { [key in AssetType]: Asset } = assetsDefinition
