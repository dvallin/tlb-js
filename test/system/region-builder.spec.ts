import { RegionBuilder } from '../../src/systems/region-builder'
import { Uniform } from '../../src/random/distributions'
import { World } from '../../src/ecs/world'
import { TlbWorld, registerComponents, registerResources } from '../../src/tlb'
import { RegionComponent } from '../../src/components/region'
import { Rectangle } from '../../src/geometry/rectangle'
import { mockRenderer } from '../mocks'
import { renderMap } from '../render'
import { Vector } from '../../src/spatial'
import { WorldMapResource } from '../../src/resources/world-map'
import { features } from '../../src/assets/features'

describe('RegionBuilder', () => {
  let world: TlbWorld
  let system: RegionBuilder
  beforeEach(() => {
    world = new World()
    registerComponents(world)
    registerResources(world, mockRenderer())
    system = new RegionBuilder(new Uniform('12'))
  })

  describe('red region', () => {
    it('creates the region', () => {
      const shape = new Rectangle(0, 0, 30, 30)
      const region = world
        .createEntity()
        .withComponent('active', {})
        .withComponent<RegionComponent>('region', {
          type: 'red',
          shape,
          level: 0,
          authorized: new Set(),
          exits: [],
        }).entity

      system.update(world, region)

      expect(world.getStorage('structure').size()).toBeGreaterThan(0)
      expect(renderMap(world, shape)).toMatchSnapshot()
    })

    it('puts corridors at exits', () => {
      const shape = new Rectangle(0, 0, 30, 30)
      const exits = [new Vector([0, 10]), new Vector([25, 29])]
      const region = world
        .createEntity()
        .withComponent('active', {})
        .withComponent<RegionComponent>('region', {
          type: 'red',
          shape,
          level: 0,
          authorized: new Set(),
          exits,
        }).entity

      system.update(world, region)

      expect(world.getStorage('structure').size()).toBeGreaterThan(0)
      expect(renderMap(world, shape)).toMatchSnapshot()

      const level = world.getResource<WorldMapResource>('map').levels[0]
      exits.forEach(e => expect(level.tileMatches(world, e, f => f!.feature() === features.corridor)).toBeTruthy())
    })
  })

  describe('elevator region', () => {
    it('creates the region', () => {
      const shape = new Rectangle(0, 0, 5, 5)
      const region = world
        .createEntity()
        .withComponent('active', {})
        .withComponent<RegionComponent>('region', {
          type: 'elevator',
          shape,
          level: 0,
          authorized: new Set(),
          exits: [],
        }).entity

      const system = new RegionBuilder(new Uniform('12'))

      system.update(world, region)

      expect(renderMap(world, shape)).toMatchSnapshot()
    })

    it('puts corridors at exits', () => {
      const shape = new Rectangle(0, 0, 5, 5)
      const exits = [new Vector([0, 2])]
      const region = world
        .createEntity()
        .withComponent('active', {})
        .withComponent<RegionComponent>('region', {
          type: 'elevator',
          shape,
          level: 0,
          authorized: new Set(),
          exits,
        }).entity

      system.update(world, region)

      expect(world.getStorage('structure').size()).toBeGreaterThan(0)
      expect(renderMap(world, shape)).toMatchSnapshot()

      const level = world.getResource<WorldMapResource>('map').levels[0]
      exits.forEach(e => expect(level.tileMatches(world, e, f => f!.feature() === features.corridor)).toBeTruthy())
    })
  })
})
