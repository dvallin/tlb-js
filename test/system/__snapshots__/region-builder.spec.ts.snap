// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegionBuilder elevator region creates the region 1`] = `
"     
 ... 
 .↕. 
 ... 
     "
`;

exports[`RegionBuilder elevator region puts corridors at exits 1`] = `
"     
.... 
..↕. 
.... 
     "
`;

exports[`RegionBuilder red region creates the region 1`] = `
"                    ..∏       
 ......... ........ ... ..... 
 ......... ........ ╔╗. ..... 
 ......... ........ ╚╝.◊..... 
 .g....... ........ ...◊..... 
 ......... ........ .╔╗◊..... 
 .........◊........ .╚╝ Â.... 
 ........◊........ ... ....g 
 .........◊........ ...  ◊◊◊  
 ......... ........ ... ..gg. 
 ......... ........ ...◊..... 
 ......... ........ ...◊..... 
 ......... ...cÂ... .╔╗◊ππ... 
 ......... ........ .╚╝ ππ.gg 
    ◊◊◊      ◊◊◊    ...  ◊◊◊  
..............................
..............................
..............................
           ...      ◊◊◊       
 ........ ... .............. 
 .....g... ...◊.............. 
 ......... ...◊.............. 
 ......... ...◊.............. 
 ......... ... .............. 
    ◊◊◊    ...      ◊◊◊       
 .........◊...◊.............. 
 .........◊...◊.............. 
 .........◊...◊.............. 
 ......... ... .............. 
           ...                "
`;

exports[`RegionBuilder red region puts corridors at exits 1`] = `
"               ...            
 ........ .... .∏. ..... .... 
 .....g.. .... ... g..π. .... 
 ........◊.... ... ..gππ◊g... 
 ........◊.... ╔╗. π...π◊.... 
 .......◊.... ╚╝. π..π.◊Â... 
 ........ .... ... ...π. .... 
 ........ .Â.c ... ..ππ. .... 
   ◊◊◊    ◊◊◊  ...  ◊◊◊  ◊◊◊  
..............................
..............................
..............................
             ...  ◊◊◊   ...◊◊◊
 ........... ... ...... ... . 
 g..........◊... ...... ... . 
 ...........◊... ...... ...◊. 
 ..........◊╔╗. ...... ...◊. 
 ........... ╚╝∏ ...... ...◊. 
             ... ...... ... . 
 ........... ... ...... ... . 
 ........... ... ...... ... . 
 ........... ... ...... ...   
 ...........◊...  ◊◊◊   ... . 
 ...........◊╔╗. ...... ... . 
 ...........◊╚╝.◊......◊...◊. 
 ........... ...◊......◊...◊. 
 ........... ...◊......◊...◊. 
 ........... ╔╗. ...... ... . 
 ........... ╚╝. ...... ... . 
             ...        ...   "
`;
